### ISSUE TITLE
How to start  background_tasks  on event  startup

### ISSUE THREAD

**Issue by HuRuWo** [tiangolo/fastapi#3214]
### First check

* [ ] I added a very descriptive title to this issue.
* [ ] I used the GitHub search to find a similar issue and didn't find it.
* [ ] I searched the FastAPI documentation, with the integrated search.
* [ ] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}
```

### Description

* Open the browser and call the endpoint `/`.
* It returns a JSON with `{"Hello": "World"}`.
* But I expected it to return `{"Hello": "Sara"}`.

### Environment

* OS: [e.g. Linux / Windows / macOS]:
* FastAPI Version [e.g. 0.3.0]:

To know the FastAPI version use:

```bash
python -c "import fastapi; print(fastapi.__version__)"
```

* Python version:

To know the Python version use:

```bash
python --version
```

**Comment by Hultner** [comment]
Where are your background_tasks defined? The problem right now is that you are expecting it as an argument for the startup_event which is called upon by the lifecycle handler without this argument. You'd have to retrieve this information from somewhere.

You might be interested in the [EasySchedule](https://github.com/codemation/easyschedule) library by @codemation which handles task scheduling in a more fine grained manner.

**Comment by tusharjagtap-bc** [comment]
Any update on this? I'm hitting the same issue.

**Comment by codemation** [comment]
@tusharjagtap-bc this is a non-issue. 

on_event('startup') does not expect the decorated method to have arguments. Ask yourself, why are arguments required in this task?

**Comment by tusharjagtap-bc** [comment]
@codemation That is an issue in my opinion. on_event('startup') does not mean that you will always want to call a method without arguments on startup, does it?

**Comment by codemation** [comment]
The most obvious way to achieve referencing your method that takes arguments, is with a new method with the arguments statically defined: 

```
@app.on_event('startup')
async def startup_event_setup():
    startup_event(background_tasks)
```

@tusharjagtap-bc It might help if you share what argument that you are passing into your startup event.

**Comment by Hultner** [comment]
As codemation said, you must have these background_tasks referenced somewhere, so calling the desired function from within the event handler makes the most sense. You could also provide it as a default argument but I don't see much reason to do that.

**Comment by stur86** [comment]
> As codemation said, you must have these background_tasks referenced somewhere, so calling the desired function from within the event handler makes the most sense. You could also provide it as a default argument but I don't see much reason to do that.

No, this is incorrect. In regular entry points like `.get`, `background_tasks` simply gets injected automatically by FastAPI. You don't have to instantiate it outside. So you can't in fact solve this problem that way. Perhaps the hack that might do it is to set up a path that launches the background task and then have an `on_startup` event that sends a request to the app itself to kick it off. But it's silly that one has to resort to such means.

**Comment by Leohanhart** [comment]
Just saying, 22-05-2022 now. Im adding a background task on startup, just the way I would expect it. 
I followed instruction, I made a static function, added a startup_event. Tried it on any way. 
    
@app.on_event("startup")
async def startup_event(background_tasks: BackgroundTasks):
    
    #tried it as startup_script.start_update_scedule()
    startup_event(background_tasks.add_task(onstart_function))
    # tried it as background_tasks.add_task(onstart_function)
 
and nothing worked. This is a serious issue here, make it !$@%! work. 
Prettending this problem is not an real problem - even if it is brought to you mulitple times - makes me rember why I never hire  !@#!% narcists. Start making clear policy's on who to start a !@#%^^& background task when the program starts.

**Comment by jayvynl** [comment]
Here is a duplicated question and this is the answer https://github.com/tiangolo/fastapi/issues/2713#issuecomment-768949823

**Comment by humblepie** [comment]
Please provide dependency injection for BackgroundTasks in startup_event to solve this issue.  Would be consistent with how the route handlers work.
